### ISSUE TITLE
TransformerDecoder.forward()

### ISSUE THREAD

**Issue by haoming-codes** [FireRedTeam/FireRedASR#59]
I wrote a `forward()` function for the `fireredasr.models.module.transformer_decoder.TransformerDecoder` in an attempt to finetune the AED model with `CrossEntropyLoss`. It appears that the loss converges but the WER does not. Will you kindly check if my implementation is on the right track? Thank you!

```
class TransformerDecoder:
    def forward(self, tgt, memory, src_mask):
        # Compute target mask for causal masking and padding
        tgt_mask = self.ignored_target_position_is_0(tgt, self.pad_id)

        # Embedding + Positional Encoding
        dec_output = self.dropout(
            self.tgt_word_emb(tgt) * self.scale +
            self.positional_encoding(tgt)
        )

        # Pass through decoder layers
        for layer in self.layer_stack:
            dec_output = layer.forward(
                dec_output, memory,
                tgt_mask, src_mask,
                cache=None
            )

        # Normalize and project to logits
        dec_output = self.layer_norm_out(dec_output)
        logits = self.tgt_word_prj(dec_output)
        return logits

class FireRedAsrAed:
    def forward(self, padded_input, input_lengths, labels=None, **kwargs):
        enc_outputs, _, enc_mask = self.firered_model.encoder(
            padded_input, 
            input_lengths)
        if labels is not None:
            decoder_input_ids = shift_tokens_right(
                labels, 
                self.config.pad_token_id,
                self.config.bos_token_id)
        lm_logits = self.firered_model.decoder(
            tgt=decoder_input_ids, 
            memory=enc_outputs, 
            src_mask=enc_mask)
        
        if labels is not None:
            loss_fct = CrossEntropyLoss()
            labels = labels.to(lm_logits.device)
            loss = loss_fct(lm_logits.view(-1, self.vocab_size), labels.reshape(-1))
        
        return Seq2SeqLMOutput(
            loss=loss,
            logits=lm_logits
)
```
